src/platform_impl/macos/window_delegate.rs

```diff
@@ -319,6 +319,7 @@ extern "C" fn window_did_become_key(this: &Object, _: Sel, _: id) {
 extern "C" fn window_did_resign_key(this: &Object, _: Sel, _: id) {
     trace!("Triggered `windowDidResignKey:`");
     with_state(this, |state| {
+        state.emit_event(WindowEvent::ModifiersChanged(ModifiersState::empty()));
```

Comment:
I've played around with things on macOS too, but haven't gotten a chance to send a PR yet. But based on my understanding, this would break when focus is lost with a modifier pressed and the modifier is still held when the window is focused again, right?

Based on my testing, there is no flags changed emitted when the window is focused or unfocused, so sending an empty on unfocus only resets things to none, but it doesn't correctly make sure that the modifier is reapplied on enter?

Reply:
That is correct—the changes I made did not re-apply the former pre-unfocus modifiers once focus is regained. The Wayland implementation does have and use a `modifiers_tracker`, but I didn't find something that was already built that I could use, besides the [`modifiers` field on `ViewState`](https://github.com/rust-windowing/winit/blob/b28fd6b4fa4ac0d1e6232d5aa6d290e3c8d8e6cb/src/platform_impl/macos/view.rs#L59). I also wasn't sure _if_ we even needed it (modifiers tracking), since in most of the cases I tested I had to release all modifiers to regain focus anyhow. (The only exception is mousing out of the window, but personally I'd still expect to have to freshly hold down a modifier if I want to use it, I guess.)

Is there a way we can figure out (e.g. by querying or looking at a field) what the current modifiers are when focus is regained? I'm not sure what the best approach here is as I have absolutely zero experience with macOS window development. What is "correct" doesn't seem to be clear to me, besides clearing modifiers when the window is unfocused.

I was primarily targeting the <kbd>⌘-TAB</kbd> use case here, so I'd definitely be open to additional ideas if you have them.

Reply:
> since in most of the cases I tested I had to release all modifiers to regain focus anyhow

This is definitely not the case. In my testing it was quite trivial to do this.

> Is there a way we can figure out (e.g. by querying or looking at a field) what the current modifiers are when focus is regained?

When I looked at it, it didn't seem like the `state` we have on the focus callback would give us access to it. I have no experience with macOS either, though I was able to write a reliably working solution that uses key events in addition to flag changes. You can check if `state.modifiers != event_mods` and then emit an event and set `state.modifiers` to `event_mods`.

That seemed to work pretty well, though I have not tested it sufficiently to determine if this solution would work reliably in all scenarios. But based on my initial estimate, it at least seemed to give good feedback for the key events.

The problem of course is that we'd have to wait until a key event to update the modifiers, which means that they are not correct until another key is pressed.

Reply:
>This is definitely not the case.

Yeah, I realized this after I wrote it. The mouse is a thing, after all. :)

As I thought about it more, I realized it'd work to do this check within `mouse_click`, `key_down`, and `key_up` within the `view` module. That check would be fairly straightforward to do, and since these all come from `NSEvent` the `modifierFlags` should definitely be available, and checkable via `event_mods` as you suggested.

I'll go ahead and PR that as well.

Reply:
> As I thought about it more, I realized it'd work to do this check within mouse_click, key_down, and key_up within the view module. That check would be fairly straightforward to do, and since these all come from NSEvent the modifierFlags should definitely be available, and checkable via event_mods as you suggested.

That's not entirely accurate though. If a modifier is released while the window is not in focus, you won't accurately be able to tell pressed modifiers until one of these actions occurs, even though your window already has full focus. So that's definitely a bit strange.

I feel like there should be a solution without caveats like this, but I haven't had the time to look more into that yet.

Reply:
@chrisduerr, yeah, that's right. The fundamental problem is that we don't seem to be able to handle a `NSEvent` when focus is regained, because such an event isn't emitted. If such an event were emitted, we could just grab the modifiers out of it.

I've done quite a bit of digging now and I'm still getting more and more confident that our answer will be to just check modifiers and queue up an event to update `state.modifiers` if necessary.

While the docs I linked above do say that

>You can retrieve the window object in question by sending `object` to `notification`,

the "window object" one would retrieve would be a `NSWindow`, which [does not expose any modifiers information](https://developer.apple.com/documentation/appkit/nswindow). As such, in the context of handling a didResignKey or didBecomeKey event, we simply _don't have access_ to the modifier information.

So I think this is a protocol-level issue: our notification of gaining or losing focus gets fired on the `NSWindowDelegate`, not on the `NSView`, where keyboard, mouse, and "flags" events get fired, and indeed no event is consistently fired after regaining focus on our `NSView` that would let us get updated modifiers. Unless I'm missing something, that'd mean we don't really have a way, besides periodically checking the provided modifiers with an event that fires and sending a fresh event before sending the event that we got.

---

I've got [another PR into this branch](https://github.com/murarth/winit/pull/5) that will add this check in a couple of key spots where I think user interaction is most likely. And of course, if `ModifiersChanged` gets fired, then `state.modifiers` will be consistent and events will not get doubly fired.

Reply:
How does this code interact with the new changes? I'd imagine it might not be necessary anymore?

What happens if a modifier is held, the window is unfocused and the modifier is still held when entering again? Does this cause an empty event without another modifier event because the internal window state never let it go?

Reply:
>I'd imagine it might not be necessary anymore?

That's an interesting question, and I think you're right that it's probably not be necessary to keep it. Anecdotally, commenting out this line did not change the fact that Alacritty is fixed by this PR.

One point of contention is that the Wayland impl _is_ emitting an empty event and then another to restore it when focus is restored. But the Wayland event lifecycle is not the same as on macOS.

>What happens if a modifier is held, the window is unfocused and the modifier is still held when entering again? Does this cause an empty event without another modifier event because the internal window state never let it go?

In this case, the window losing focus triggers an emptying event, and regaining focus does not trigger an event until selected further user interaction happens. (Either keypress, mouse click, mouse enter, mouse exit, scrolling, or pressure change.) You're right that we'd be firing an empty event. I suppose since we now have a check before actually sending out events for user interaction, we no longer need to send out empty modifiers state, since windows might not want to get a synthetic event like that.

I should also do this check (potentially updating stale modifiers) on `mouse_motion`, I think, since it's pretty lightweight.